x11rb-async/src/rust_connection/mod.rs

```diff
@@ -0,0 +1,488 @@
+                let res = CURRENT_CONTEXT.set(cx.waker(), || f(self));
+                Poll::Ready(res)
+            })
+            .await;
+
+            match res {
+                Err(e) if e.as_io_error().map(|e| e.kind()) == Some(io::ErrorKind::WouldBlock) => {}
+                res => return res,
+            }
+
+            // Wait until the stream is readable or writable.
+            match mode {
+                PollMode::Readable => self
+                    .inner
+                    .stream()
+                    .stream
+                    .readable()
+                    .await
+                    .map_err(E::from_io_error)?,
+                PollMode::Writable => self
+                    .inner
+                    .stream()
+                    .stream
+                    .writable()
+                    .await
+                    .map_err(E::from_io_error)?,
+                PollMode::ReadAndWritable => {
+                    // `or` them together.
+                    let read = self.inner.stream().stream.readable();
+                    let write = self.inner.stream().stream.writable();
+
+                    read.or(write).await.map_err(E::from_io_error)?;
+                }
+            }
+        }
+    }
+}
+
+impl<S: Stream + Send + Sync + 'static> RequestConnection for RustConnection<S> {
+    type Buf = Vec<u8>;
+
+    fn send_request_with_reply<'this, 'bufs, 'sl, 'future, R>(
+        &'this self,
+        _bufs: &'bufs [io::IoSlice<'sl>],
+        _fds: Vec<RawFdContainer>,
+    ) -> Fut<'future, Cookie<'this, Self, R>, ConnectionError>
+    where
+        'this: 'future,
+        'bufs: 'future,
+        'sl: 'future,
+        R: x11rb::x11_utils::TryParse,
+    {
+        Box::pin(async move {
+            // Gain exclusive access to the write end of the stream.
+            let _guard = self.write_lock.lock().await;
+
+            todo!()
+        })
+    }
+
+    fn send_request_with_reply_with_fds<'this, 'bufs, 'sl, 'future, R>(
+        &'this self,
+        _bufs: &'bufs [io::IoSlice<'sl>],
+        _fds: Vec<RawFdContainer>,
+    ) -> Fut<'future, CookieWithFds<'this, Self, R>, ConnectionError>
+    where
+        'this: 'future,
+        'bufs: 'future,
+        'sl: 'future,
+        R: x11rb::x11_utils::TryParseFd,
+    {
+        todo!()
+    }
+
+    fn send_request_without_reply<'this, 'bufs, 'sl, 'future>(
+        &'this self,
+        _bufs: &'bufs [io::IoSlice<'sl>],
+        _fds: Vec<RawFdContainer>,
+    ) -> Fut<'future, VoidCookie<'this, Self>, ConnectionError>
+    where
+        'this: 'future,
+        'bufs: 'future,
+        'sl: 'future,
+    {
+        todo!()
+    }
+
+    fn check_for_raw_error(
+        &self,
+        sequence: SequenceNumber,
+    ) -> Fut<'_, Option<Self::Buf>, ConnectionError> {
+        Box::pin(async move {
+            // Gain exclusive access to the read end of the stream.
+            let _guard = self.read_lock.lock().await;
```

Comment:
What if the `Error` was already received from the server (so sits in some queue somewhere), but another task sits in `wait_for_event()` and thus this mutex is locked? That would produce unwanted blocking.